object-versioning/bucket-level-versioning-configuration.md

```diff
@@ -0,0 +1,150 @@
+
+### Extensibility
+The design should support or not deter additional bucket level configuration and/or constraints in the metadata.
+
+### Customizable defaults
+Support for customizable defaults that can be defined and configured at the project/account level.
+
+### Minimal impact
+The operational impact to the system with the addition of this bucket level configuration should be limited to the
+API calls of version based operations only.
+
+These API calls include
+* `GET Bucket versioning`
+* `PUT Bucket versioning`
+* `GET Object`
+* `PUT Object`
+* `DELETE Object`
+* `List Object Versions`
+* `Copy Object`
+* `Upload Part Copy`
+* `Restore Object`
+
+
+## Design and implementation
+In AWS S3, Versioning is a per-bucket subresource, with 3 possible states. By default, newly created buckets are in 
+the `unversioned` state. Once versioning for a bucket is `enabled` its state can only be changed to `suspended` or 
+re-`enabled`.
+
+To support AWS s3 compatability, we will add a `versioning_state` field to the `bucket_metainfo` table. The state of 
+this field will be consulted during certain operations to allow for the correct versioning behavior. Unlike the AWS 
+versioning subresource, the `versioning_state` field will have 4 possible states, `unversioned`, `enabled`, 
+`suspended`, and a new state called `unsupported`. The first three states all behave in the same matter as the AWS 
+states. The new `unsupported` state will be used to indicate that the bucket does not support versioning, and will 
+be the state for all buckets that exist prior to the completion of the object versioning implementation.
+
+To support customizable defaults, we will add a `default_versioning_state` field to the `project` table. The state of
+this field will be consulted during bucket creation to set the initial state of the `versioning_state` field for the
+bucket. The default value for this field will be `unversioned`.
+
+### Database Schema
+#### Bucket Metainfo
+The `versioning_state` field will be added to the `bucket_metainfo` table.
+
+```
+model bucket_metainfo (
+    ...
+    field versioning_state    int (nullable, updatable)
+)
+```
+
+This int type will have 4 possible values: `null` which represents a bucket for which versioning is not supported. Any 
+bucket with a versioning state of null cannot transition to any other versioning state. A value of 0 (`unversioned`),
+represents a bucket which is eligible for object versioning, but has never had it enabled. A value of 1 (`enabled`) 
+represents a bucket which has versioning enabled, and a value of 2 (`suspended`) represents a bucket which has 
+versioning suspended. By default, new buckets will have a versioning state of 0 (`unversioned`), however, the 
+default value can be overridden by the `default_versioning_state` field in the `project` table.
+
+| Value  | Description          |
+|--------|----------------------|
+| null   | unsupported          |

```

Comment:
Any reason not to use:

```
default 0; unsupported
1; unversioned
2; enabled
3; suspended
```

It feels like pointer handling adds more complexity.